Library/Homebrew/livecheck/strategy/github_latest.rb

```diff
@@ -141,7 +141,11 @@ def self.find_versions(url:, regex: DEFAULT_REGEX, **_unused, &block)
 
           match_data[:url] = generated[:url]
 
-          release = GitHub.get_latest_release(generated[:username], generated[:repository])
+          release = GitHub::API.open_rest(
```

Comment:
Yep, it's to avoid having to add all the same keyword parameters to `#get_latest_release` and pass them into the method's `#open_rest` call. As it stands, that method is a very slim wrapper around `#open_rest`, where all it does is create a URL from the `username` and `repository` arguments before calling `#open_rest` ([we already do this in `GithubLatest#generate_input_values`](https://github.com/Homebrew/brew/blob/15244d8a7d67ed68cb02c24e9f5c4cf15b9f3d70/Library/Homebrew/livecheck/strategy/github_latest.rb#L75)):

```ruby
def self.get_release(user, repo, tag)
  url = "#{API_URL}/repos/#{user}/#{repo}/releases/tags/#{tag}"
  API.open_rest(url, request_method: :GET)
end
```

Assuming we won't use `**options`, the required changes would turn the method into something like this:

```ruby
def self.get_release(
  user,
  repo,
  tag,
  debug: nil,
  verbose: nil,
  timeout: nil,
  connect_timeout: nil,
  max_time: nil,
  retries: nil
)
  url = "#{API_URL}/repos/#{user}/#{repo}/releases/tags/#{tag}"

  curl_options = {
    debug:           debug,
    verbose:         verbose,
    timeout:         timeout,
    connect_timeout: connect_timeout,
    max_time:        max_time,
    retries:         retries,
  }.compact
  API.open_rest(url, request_method: :GET, **curl_options)
end
```

It's worth mentioning that this is the method _without_ a type signature, so it will cover even more lines if/when this is eventually typed.

The method call in `GithubLatest#find_versions` would be:

```ruby
release = GitHub.get_latest_release(
  generated[:username],
  generated[:repository]),
  **Strategy::DEFAULT_CURL_OPTIONS,
)
```

The only difference is that we would be calling a `GitHub` method and the arguments would be different (it's the same number either way).

As mentioned, we already generate the relevant API URL in `GithubLatest`, so we don't benefit from having `#get_latest_version` generate the URL for us. We would only really benefit from it if the API URL will change in the future (i.e., `GithubLatest` would automatically get any fix in `GitHub`) but that seems pretty unlikely and fixing it in `GithubLatest` is very easy. We can get most of that benefit by using `GitHub::API_URL` in the generated URL string in `GithubLatest` but I'll address that in another PR.

[For what it's worth, we generate the URL in the strategy because we need to print it in debug runs and include it in verbose JSON output, so it's not something that we can omit from the strategy and we can't get the URL that was used back from `#open_rest`/`#github_latest_release`.]

Past that, this would be the only `GitHub` method that supports options and, as mentioned, it would only support the ones that we've explicitly added here. If we ever needed more, we would have to modify `#open_rest` and update any affected `GitHub` methods. So, we would end up with a mishmash of option support that's a challenge to maintain and reason about.

So, while using `GitHub#get_latest_release` may feel a little nicer on some level, everything we would need to do to make that happen feels decidedly not nice. Using `#open_rest` allows us to avoid making more of a mess and we don't lose anything meaningful in the process. [The proposed `GithubReleases` strategy also uses an `#open_rest` call in its `#find_versions` method, so using `#open_rest` in `GithubLatest` isn't that different.]

Reply:
> the required changes would turn the method into something like this:

Maybe a stupid question: why does Livecheck need all of these options?

> Past that, this would be the only `GitHub` method that supports curl options and it would only support the ones that we've explicitly added here.

This seems better to me. I also think it's worth adding typing as you go along here.

Reply:
> Maybe a stupid question: why does Livecheck need all of these options?

Always happy to explain if/when it's helpful. Some of the options are [explained in comments above the related `Strategy` constants](https://github.com/Homebrew/brew/blob/5f2ccd22f2ed3b90ab4ebdc52c9aba2b149de1b5/Library/Homebrew/livecheck/strategy.rb#L23-L36) but I'll go over them all here, for the sake of completion:

* `debug: false`: `brew livecheck --debug` output is carefully crafted but curl methods print additional output for debugging when the `--debug` flag is present. Without `debug: false`, we would see the `curl` command in the middle of the debug output when we run `brew livecheck --debug` on a check that uses curl underneath. This is what's currently happening for `GithubLatest` checks when `--debug` is used and is part of the impetus for this PR.
* `verbose: false`: I believe I originally added `verbose` mostly as a precaution alongside `debug`, with a similar rationale. Basically, we use `--verbose` for `--debug` and `--json` runs to include more information in the output and I didn't want the flag to cause `curl` to print output that we don't want/need (or _more_ than we want/need). However, I did some testing with `verbose` removed and didn't see any difference, so I'll update this to drop that option (and we can always add it back if/when there's a confirmed need for it).
* `connect_timeout: CURL_CONNECT_TIMEOUT` (10): The default `--connect-timeout` value in curl (the application) can be up to two minutes, so we need to set a more reasonable duration to avoid a lengthy wait when a connection can't be established.
* `max_time: CURL_MAX_TIME` (15): curl (the application) doesn't have a default `--max-time` value, so we set a value to ensure curl will time out in a reasonable amount of time.
* `timeout: CURL_PROCESS_TIMEOUT` (20): The `curl` process will sometimes hang indefinitely (despite setting the `--max-time` argument) and it needs to be quit for livecheck to continue.
* `retries: 0`: Retrying a failed request makes sense in critical parts of `brew` (e.g., downloading a file for installation, etc.) but it's arguably not as useful for livecheck, in my experience. Some servers that livecheck checks are a little flaky and will randomly/periodically time out but an immediate retry is pretty much guaranteed to fail (though they sometimes work if you wait a while to retry or simply get lucky). With that in mind, my thinking is that retrying failed requests in livecheck would typically just be a waste of effort (i.e., unnecessarily using resources on the host and remote server), so we don't bother because we can just run the check again later. If it's a long-term issue, we simply modify the check. If you need evidence to either support or oppose this position I can try to gather data but it's a bit tricky because this retry would happen within curl itself, so we would only be able to see whether retrying potentially allows a check to work that otherwise may not. To get a better sense, I would have to do something like implement some temporary retry logic in `Utils::Curl` that we can keep track of and check those statistics. That would probably be left to a follow-up PR, though.

> This seems better to me. I also think it's worth adding typing as you go along here.

Modifying `#get_latest_release` doesn't feel like a worthwhile tradeoff to me but I've already explained that in detail (and outlined the benefits/detriments), so I'll go ahead and update this accordingly. In the grand scheme of things, this is [theoretically] only a temporary situation, as the added parameters will be replaced if/when we improve the args/options situation in `Utils::Curl` methods in the future.

Reply:
> This feels like a smell to me for this method to behave consistently except when used by livecheck? I think livecheck output should be consistent with the rest of Homebrew rather than have special-case exceptions here.

While I understand the general desire for consistency, we do this because the curl debug text wouldn't provide any meaningful benefit for livecheck debugging probably 99+% of the time and it appears right in the middle of livecheck's debug output. For example, this is what we currently see when running `brew livecheck --debug` with a `GithubLatest` check:

```
/opt/homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /opt/homebrew/Library/Taps/homebrew/homebrew-core/Formula/adios2.rb

Formula:          adios2
Livecheckable?:   Yes

URL (stable):     https://github.com/ornladios/ADIOS2/archive/v2.9.0.tar.gz
Strategy:         GithubLatest
/usr/bin/env /opt/homebrew/Library/Homebrew/shims/shared/curl --disable --cookie /dev/null --globoff --show-error --user-agent Homebrew/4.0.17-65-g082017e\ \(Macintosh\;\ arm64\ Mac\ OS\ X\ 13.3.1\)\ curl/7.87.0 --header Accept-Language:\ en --retry 3 --location https://api.github.com/repos/ornladios/ADIOS2/releases/latest --header Accept:\ application/vnd.github\+json --write-out '
'\%\{http_code\} --header Authorization:\ token\ ****** --header X-GitHub-Api-Version:2022-11-28 --dump-header /private/tmp/github_api_headers20230511-56768-1ig1ga4
URL (strategy):   https://api.github.com/repos/ornladios/ADIOS2/releases/latest
Regex (strategy): /v?(\d+(?:\.\d+)+)/i

Matched Versions:
2.9.0

adios2: 2.9.0 ==> 2.9.0
```

The desired output is:

```
/opt/homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /opt/homebrew/Library/Taps/homebrew/homebrew-core/Formula/adios2.rb

Formula:          adios2
Livecheckable?:   Yes

URL (stable):     https://github.com/ornladios/ADIOS2/archive/v2.9.0.tar.gz
Strategy:         GithubLatest
URL (strategy):   https://api.github.com/repos/ornladios/ADIOS2/releases/latest
Regex (strategy): /v?(\d+(?:\.\d+)+)/i

Matched Versions:
2.9.0

adios2: 2.9.0 ==> 2.9.0
```

There have been a handful of times over the years where seeing the raw curl command is a _little_ helpful (though not essential) but it's not worth having at the expense of livecheck's default debug output. I could maybe understand having it available for `--debug --verbose` (where we already adapt the debug output to include a little more information, like printing the underlying `Version` objects for matched versions instead of only the text). However, it would come at the expense of readability, one way or another.

Since we can't control when/how debug output is printed in `Utils::Curl` (e.g., we can't collect that output and print it in a more appropriate location), the only way to make that possible with the current setup would be to not print any output for a given check until after it completes. In that situation, the curl output would appear before the "URL", "Strategy", etc. output.

That's a poor experience for users because they wouldn't know what check is being run until after it's done. The curl output would provide an indication that the check is being run but it's very dense and identifying the URL in that text is a chore. Additionally, that would only be the case for checks that use curl. Some checks (e.g., `Git`) would show nothing until after the check completes.

The reason why we don't do this is because printing information as soon as it's available provides feedback to the user that livecheck is working (and curl output is just noise the vast majority of the time). It's technically possible to print information as we go while also printing the curl output but it would require a fundamental reworking of the debug output and I think it would end up being more verbose and take more effort to visually parse compared to the current formatting.

-----

> These all seem like things where our current `curl` usage can have better defaults rather than special-casing livecheck here.

What makes that tricky is `Utils::Curl` has to support a variety of use cases, which sometimes have opposing requirements. We may be able to have somewhat tighter defaults in `Curl` methods but they would likely be higher than the values that we're using in livecheck.

Kind of like the retry situation, having longers timeouts/max values may make sense for critical parts of brew (installation, etc.), where it's preferable to wait and potentially succeed rather than fail in a shorter amount of time. The values for livecheck are oriented towards not having any single check hold up the run for too long (a homebrew/core tap run already takes around an hour as it is).

So, while I agree that there may be some value in investigating whether we can have better details in `Utils::Curl`, livecheck will almost certainly still have to set different values that are appropriate for its specific context. That said, I'm not familiar enough with other parts of brew to say what would be good default values, so it would probably be good to create an issue and ping Homebrew/brew for ideas if we want to pursue it outside of this PR.

Reply:
> For example, this is what we currently see when running brew livecheck --debug with a GithubLatest check:

I think it's desirable/better to see the `curl` call there, albeit with better whitespace usage.

> That's a poor experience for users because they wouldn't know what check is being run until after it's done. 

Users aren't using `--debug` for typical usage. Non-contributor users aren't really using `livecheck` for that matter, either.

> and curl output is just noise the vast majority of the time

This is the case of most debug output most of the time. It's not useful until it is.

> We may be able to have somewhat tighter defaults in Curl methods but they would likely be higher than the values that we're using in livecheck.

This seems fine.

> a homebrew/core tap run already takes around an hour as it is

I don't think this is worth optimising for. It's unclear that anyone except you is doing this.

> I'm not familiar enough with other parts of brew to say what would be good default values, so it would probably be good to create an issue and ping Homebrew/brew for ideas if we want to pursue it outside of this PR.

I am familiar: let's talk about it here and bring in @Homebrew/brew for more thoughts.